api/docs/scatter_gather_emulation.dox

```diff
@@ -0,0 +1,230 @@
+
+
+## Drreg Support For Multi-phase Reservations
+
+Owner: [Abhinav Sharma](https://github.com/abhinav92003)
+
+Upstream issue: [DynamoRIO/dynamorio#3823](https://github.com/DynamoRIO/dynamorio/issues/3823)
+
+Drreg is DynamoRIO’s register reservation framework. It allows users to reserve a reg to use as scratch. Internally, drreg automatically performs the following functions so that the user does not need to. Drreg
+- keeps all required book-keeping like the spill slot to spilled register mapping
+- restores spilled registers to their application value before they are read by an application instruction; also, it re-spills the spilled registers if they are written by an application instruction.
+- performs application state restoration on state restore events like encountering an application fault, and DR detach.
+
+While expanding a scatter or gather instruction in the app2app phase, we need a scratch register to hold the scalar values and masks. In later phases (like the insertion or the instru2instru phase), drcachesim and other DR clients may also use drreg to get scratch registers for their instrumentation.
+
+Drreg initially supported only single phase use and did not keep any state between phases. To support multi-phase use of drreg, we need to solve the following:
+- avoid spill slot conflict across multiple phases: multi-phase use can potentially lead to spill slot conflicts if the same slot is selected in multiple phases. This may clobber the spilled application value and cause the application to crash or otherwise fail.
+- allow aflags spill to any slot: drreg hardcodes the aflags spill slot as the zero-th slot, to simplify some logic. To support the ability to spill aflags in multiple phases, drreg should be able to use any spill slot for aflags.
+- application state restore logic: on a state restore event, we should be able to figure out which slot contains each spilled register's app value. This is complicated by the fact that registers may be spilled by instrumentation added by multiple phases, and the spill regions may overlap which causes the spilled application value to be moved between spill slots.
+
+We explored the following ideas to avoid spill slot conflicts in drreg:
+
+#### Disjoint slot spaces or arenas
+We can ask drreg to create slot spaces or arenas at init time, which are assigned disjoint spill slots. When reserving a register, the user passes in a “space/arena Id '' to instruct drreg to pick free slots only from that arena. This requires keeping some global drreg state. This also requires the user to guess the best configuration for assigning slots to the arenas, and passing the correct arena Id before each reservation. It may artifically make some spill slots unavailable for use, thereby reducing efficiency.
```

Comment:
Are those quotes right: ` “space/arena Id '' `

Reply:
Ah. Copying from docs made it weird. Fixed.